pkg/controllers/user/template/template.go

```diff
@@ -0,0 +1,135 @@
+package template
+
+import (
+	"fmt"
+
+	v1beta12 "github.com/rancher/types/apis/extensions/v1beta1"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	"github.com/rancher/types/config"
+	"github.com/sirupsen/logrus"
+	"k8s.io/api/extensions/v1beta1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+const (
+	podSecurityAnnotation        = "serviceaccount.cluster.cattle.io/pod-security"
+	podSecurityVersionAnnotation = "serviceaccount.cluster.cattle.io/pod-security-version"
+	apiVersion                   = "extensions/v1beta1"
+)
+
+func Register(context *config.UserContext) {
+	m := &manager{
+		/*clusters: context.Management.Management.Clusters(""),
+		bindings: context.RBAC.RoleBindings(""),
+		roles:    context.RBAC.Roles(""),
+
+		clusterLister:   context.Management.Management.Clusters("").Controller().Lister(),
+		templateLister:  context.Management.Management.PodSecurityPolicyTemplates("").Controller().Lister(),
+		bindingLister:   context.RBAC.RoleBindings("").Controller().Lister(),
+		roleLister:      context.RBAC.Roles("").Controller().Lister(),
+		namespaceLister: context.Core.Namespaces("").Controller().Lister(),
+		projectLister: context.Management.Management.Projects("").Controller().Lister(),*/
+		policies:     context.Extensions.PodSecurityPolicies(""),
+		policyLister: context.Extensions.PodSecurityPolicies("").Controller().Lister(),
+	}
+
+	context.Management.Management.PodSecurityPolicyTemplates("").AddHandler(
+		"PodSecurityPolicyTemplateSyncHandler",
+		func(key string, obj *v3.PodSecurityPolicyTemplate) error {
+			// find psp that use this template
+			policy, err := m.policyLister.Get("", obj.Name)
+			if err != nil {
+				if errors.IsNotFound(err) {
+					_, err := PodSecurityPolicyFromTemplate(m.policies, obj)
+					if err != nil {
+						return err
+					}
+				} else {
+					return fmt.Errorf("error getting policies: %v", err)
+				}
+			}
+
+			// check if version matches
+			if policy.Annotations[podSecurityVersionAnnotation] != obj.ResourceVersion {
+				_, err := PodSecurityPolicyFromTemplate(m.policies, obj)
+				if err != nil {
+					return err
+				}
+			}
+
+			return nil
+		})
+}
+
+type manager struct {
+	policies     v1beta12.PodSecurityPolicyInterface
+	policyLister v1beta12.PodSecurityPolicyLister
+}
+
+func PodSecurityPolicyFromTemplate(policies v1beta12.PodSecurityPolicyInterface, originalTemplate *v3.PodSecurityPolicyTemplate) (*v1beta1.PodSecurityPolicy, error) {
+	template := originalTemplate.DeepCopy()
+	template.Kind = "PodSecurityPolicy"
+	template.APIVersion = apiVersion
+	template.ResourceVersion = ""
```

Comment:
There is no reason to set the fields on template, is there?

Reply:
There is not a reason, this got missed when I was refactoring this to use deep copy.